Supports automotive bsta1000b-fada-aarch64 car platform

elliott10 opened the pull request:
* Added bsta1000b-fada-aarch64 car platform for automotive;
* Successfully initializes into apps/fs/shell program;
* Implemented functions of DW8250 uart driver, including changing the baudrate;
* Implemented the translation of GIC interrupt ID: PPI, SPI;
* Kernel and devices memory pages are mapped;
* Added use-ramdis feature for apps/fs/shell;
* Created the kernel uImage image for car platform;
* Supports 8 SMP for bsta1000b fada board

Build commands:

```
make A=apps/fs/shell PLATFORM=aarch64-bsta1000b LOG=info SMP=8 FEATURES="driver-ramdisk,irq" fada
```

equation314 commented on modules/axhal/src/platform/bsta1000b_fada_aarch64/dw_apb_uart.rs:
Put the register definition into crates (like crates/arm_pl011)

elliott10 replied:
I've put it into crates/dw_apb_uart

equation314 commented on modules/axhal/src/platform/bsta1000b_fada_aarch64/utils.rs:
```diff
@@ -0,0 +1,67 @@
+}
+
+/// 毫秒(ms) 延时器
+pub fn msdelay(ms: u64) {
```
Please use `axhal::time::busy_wait` directly

elliott10 replied:
Okay

equation314 commented on modules/axhal/src/platform/bsta1000b_fada_aarch64/utils.rs:
```diff
@@ -0,0 +1,67 @@
+use crate::mem::phys_to_virt;
+
+/// 微秒(us) 延时器
+pub fn usdelay(us: u64) {
```
Remove this function, use `axhal::time::busy_wait` directly

elliott10 replied:
Okay

equation314 commented on modules/axhal/src/platform/bsta1000b_fada_aarch64/utils.rs:
Put these functions into `misc.rs` and override `aarch64_common/misc.rs`

elliott10 replied:
These functions are only for bsta1000b board. So I put them into aarch64_bsta1000b/misc.rs

equation314 commented on scripts/bsta1000b-fada/bsta1000b-fada-arceos.its:
Move this two files into the `tools` directory.

elliott10 replied:
Okay

equation314 commented on modules/axconfig/Cargo.toml:
```diff
@@ -14,6 +14,7 @@ platform-pc-x86 = []
 platform-qemu-virt-riscv = []
 platform-qemu-virt-aarch64 = []
 platform-raspi4-aarch64 = []
+platform-bsta1000b-fada-aarch64 = []
```
The platform name is too long, can it be simplified? Such as `platform-bsta1000b`

elliott10 replied:
all right. It named aarch64-bsta1000b

equation314 commented on apps/fs/shell/Cargo.toml:
```diff
@@ -8,6 +8,10 @@ authors = ["Yuekai Jia <equation618@gmail.com>"]
 use_ramfs = ["axfs/myfs", "dep:axfs_vfs", "dep:axfs_ramfs", "dep:crate_interface"]
+
+# if needs irq, add "libax/irq"
+use-ramdisk = ["use_ramfs", "libax/use-ramdisk"]
```
It is no need to enable `use_ramfs` when enabe `use-ram disk`.

elliott10 replied:
Okay

equation314 commented on scripts/make/raspi4.mk:
```diff
@@ -28,8 +28,7 @@ ifeq ($(BSP),rpi4)
 
-# Export for build.rs.
-export LD_SCRIPT_PATH
+export RUSTFLAGS
```
It's no longer needed, as the main branch as been updated.

equation314 commented on scripts/make/bsta1000b-fada.mk:
```diff
@@ -0,0 +1,6 @@
+export RUSTFLAGS
```
It's no longer needed, as the main branch as been updated.

equation314 commented on modules/axhal/src/platform/aarch64_common/gic.rs:
```diff
@@ -37,7 +64,10 @@ pub fn register_handler(irq_num: usize, handler: IrqHandler) -> bool {
 pub fn dispatch_irq(_unused: usize) {
-    GICC.handle_irq(|irq_num| crate::irq::dispatch_irq_common(irq_num as _));
+    GICC.handle_irq(|irq_num| {
+        trace!("GICC handle irq {}", irq_num);
```
The debug info has already been printed in `src/irq.rs`

equation314 commented on modules/axhal/src/platform/aarch64_bsta1000b/mp.rs:
```diff
@@ -0,0 +1,16 @@
+    }
+    let entry = virt_to_phys(VirtAddr::from(_start_secondary as usize));
+    crate::platform::aarch64_common::psci::psci_cpu_on(
+        CPU_HWID[cpu_id],
```
Show error message if `cpu_id` is larger than 8.

equation314 commented on modules/axhal/src/platform/aarch64_bsta1000b/dw_apb_uart.rs:
```diff
@@ -0,0 +1,58 @@
+}
+
+/// Memory Barrier
+fn dmb() {
```
Use [`aarch64_cpu::asm::barrier::dmb`](aarch64_cpu::asm::barrier::dmb) instead.

equation314 commented on crates/dw_apb_uart/src/dw_apb_uart.rs:
```diff
@@ -0,0 +1,116 @@
+//! snps,dw-apb-uart serial driver
```
Move this file into `lib.rs`

equation314 commented on crates/dw_apb_uart/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+name = "dw_apb_uart"
+authors = ["Luoyuan Xiao <xiaoluoyuan@163.com>"]
+version = "0.1.0"
+edition = "2021"
```
Add more fields: description, license, homepage, repository, and documentation.

equation314 commented on apps/fs/shell/Cargo.toml:
```diff
@@ -8,6 +8,7 @@ authors = ["Yuekai Jia <equation618@gmail.com>"]
 
 [features]
 use-ramfs = ["axstd/myfs", "dep:axfs_vfs", "dep:axfs_ramfs", "dep:crate_interface"]
+use-ramdisk = ["axstd/use-ramdisk"]
```
This feature can be enabled by the `APP_FEATURES` variable when run `make`

equation314 commented on modules/axhal/src/platform/aarch64_bsta1000b/dw_apb_uart.rs:
```diff
@@ -0,0 +1,58 @@
+    UART.lock().set_ier(true);
+    dmb();
+
+    #[cfg(feature = "irq")]
```
Apply the `cfg` to the entire function `init_irq`.

equation314 commented on platforms/aarch64-bsta1000b.toml:
```diff
@@ -0,0 +1,58 @@
+# CPU_HWID = ["0x00", "0x100", "0x200", "0x300", "0x400", "0x500", "0x600", "0x700"]
+
+# BST A1000B board registers
+CPU_CSR_BASE = "0x32011000"
```
Move these values into `modules/axhal/src/platform/aarch64_bsta1000b/misc.rs`

equation314 commented on platforms/aarch64-bsta1000b.toml:
```diff
@@ -0,0 +1,58 @@
+gicc-paddr = "0x32002000"
+
+# CPU HWID from cpu device tree nodes with "reg" property
+# CPU_HWID = ["0x00", "0x100", "0x200", "0x300", "0x400", "0x500", "0x600", "0x700"]
```
Remove these comments

equation314 commented on crates/dw_apb_uart/src/lib.rs:
```diff
@@ -0,0 +1,4 @@
+//! Definitions for snps,dw-apb-uart serial
```
Add more detailed description.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
Many of `psci_*` functions are not used?

elliott10 replied:
These are reserved according to linux arch/arm64 psci interfaces.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -37,6 +34,121 @@ pub fn system_off() -> ! {
+/// PSCI v0.1
+/// cpu entry_point = phys_addr(secondary_entry)
+/// eg: psci_0_1_cpu_on, fn_id: 0xc4000003, cpu on core 256, secondary_entry: 0x80cb51f8
+pub fn psci_cpu_on(cpuid: usize, entry_point: usize, arg: usize) -> i32 {
```
Is this function the same as `cpu_on`? Keep only one.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -4,10 +4,7 @@
 
 use core::arch::asm;
 
-const PSCI_CPU_ON: u32 = 0x8400_0003;
```
PSCI codes should be the same on all platforms, remove them in the platform toml files, and define here:

```rust
/// PSCI v0.2 interface
const fn psci_0_2_fn(func: u32, is_64bit: bool) -> u32 {
    if is_64bit {
        0xc4000000 + func
    } else {
        0x84000000 + func
    }
}

const PSCI_0_2_FN_CPU_SUSPEND: u32 = psci_0_2_fn(1, false);
const PSCI_0_2_FN_CPU_OFF: u32 = psci_0_2_fn(2, false);
const PSCI_0_2_FN_CPU_ON: u32 = psci_0_2_fn(3, false);
const PSCI_0_2_FN_SYSTEM_OFF: u32 = psci_0_2_fn(8, false);
const PSCI_0_2_FN_SYSTEM_RESET: u32 = psci_0_2_fn(9, false);

const PSCI_0_2_FN64_CPU_SUSPEND: u32 = psci_0_2_fn(1, true);
const PSCI_0_2_FN64_CPU_ON: u32 = psci_0_2_fn(3, true);
```

elliott10 replied:
All right. I'll research the definition of psci address.

equation314 replied:
You can found these definitions at `/usr/include/linux/psci.h`

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
Add a function `psci_call`, and use `hvc` or `smc` respectively according to the `psci-method` field defined in the  platform toml file.

equation314 commented on modules/axhal/src/platform/aarch64_qemu_virt/mp.rs:
```diff
@@ -6,5 +6,10 @@ pub fn start_secondary_cpu(cpu_id: usize, stack_top: PhysAddr) {
     let entry = virt_to_phys(VirtAddr::from(_start_secondary as usize));
-    crate::platform::aarch64_common::psci::cpu_on(cpu_id, entry.as_usize(), stack_top.as_usize());
+    crate::platform::aarch64_common::psci::psci_cpu_on(
+        false,
```
PSCI cpu on code for QEMU virt arm should be `0xc4000003` (64bit), the origin code has a bug.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -21,22 +46,128 @@ fn psci_hvc_call(func: u32, arg0: usize, arg1: usize, arg2: usize) -> usize {
+///
+/// PSCI v0.1
+/// cpu entry_point: phys_addr(secondary_entry)
+pub fn psci_cpu_on(is_fn64: bool, cpuid: usize, entry_point: usize, arg: usize) -> i32 {
```
The `is_fn64` argument can be removed, since `PSCI_0_2_FN64_CPU_ON` is used in all cases.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -21,22 +46,128 @@ fn psci_hvc_call(func: u32, arg0: usize, arg1: usize, arg2: usize) -> usize {
+    psci_call(fn_id, arg0, arg1, arg2)
+}
+
+pub fn psci_0_1_cpu_on(fn_ids: u32, cpuid: usize, entry_point: usize, arg: usize) -> i32 {
```
The `psci_0_1_*` functions are not used and can be removed.

equation314 replied:
This function is confused with `psci_cpu_on`, keep only one.

equation314 commented on crates/dw_apb_uart/src/lib.rs:
```diff
@@ -0,0 +1,5 @@
+//! Uart snps,dw-apb-uart driver in Rust for BST A1000b FADA board.
+#![no_std]
+
+pub mod dw_apb_uart;
```
It is no need to define a new mod `dw_apb_uart`, move codes from `dw_apb_uart.rs` to `lib.rs`.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -21,22 +46,123 @@ fn psci_hvc_call(func: u32, arg0: usize, arg1: usize, arg2: usize) -> usize {
     ret
 }
 
+fn psci_call(fn_id: usize, arg0: usize, arg1: usize, arg2: usize) -> usize {
```
Returns `Result<(), PsciRetNo>`

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -21,22 +46,123 @@ fn psci_hvc_call(func: u32, arg0: usize, arg1: usize, arg2: usize) -> usize {
+/// PSCI return values, inclusive of all PSCI versions
+#[allow(non_camel_case_types)]
+#[derive(Debug)]
+pub enum PsciRetNo {
```
It's better to use the name `PsciError` to indicate an error type.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -21,22 +46,123 @@ fn psci_hvc_call(func: u32, arg0: usize, arg1: usize, arg2: usize) -> usize {
+#[allow(non_camel_case_types)]
+#[derive(Debug)]
+pub enum PsciRetNo {
+    PSCI_RET_SUCCESS = 0,
```
Returning `PSCI_RET_SUCCESS` indicates success, and it should not be a member of the error type.

equation314 commented on modules/axhal/src/platform/aarch64_common/psci.rs:
```diff
@@ -21,22 +46,123 @@ fn psci_hvc_call(func: u32, arg0: usize, arg1: usize, arg2: usize) -> usize {
+}
+
+/// PSCI return values, inclusive of all PSCI versions
+#[allow(non_camel_case_types)]
```
`PSCI_RET_INVALID_PARAMS` => `InvalidParams`.

And `#[allow(non_camel_case_types)]` can be removed.

elliott10 committed:
Supports car automotive platform bsta1000b
Build cmd: make A=apps/fs/shell ARCH=aarch64 SMP=8 PLATFORM=aarch64-bsta1000b APP_FEATURES="axstd/driver-ramdisk,axstd/irq" fada
